workermanager.go

```diff
@@ -0,0 +1,85 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/taskcluster/generic-worker/gwconfig"
+	"github.com/taskcluster/taskcluster-client-go/tcworkermanager"
+)
+
+type WorkerManagerUserData struct {
+	WorkerPoolID string          `json:"workerPoolId"`
+	ProviderID   string          `json:"providerId"`
+	WorkerGroup  string          `json:"workerGroup"`
+	RootURL      string          `json:"rootUrl"`
+	WorkerConfig BootstrapConfig `json:"workerConfig"`
+}
+
+func (userData *WorkerManagerUserData) updateConfig(c *gwconfig.Config, providerType interface{}) error {
+	wp := strings.SplitN(userData.WorkerPoolID, "/", -1)
+	if len(wp) != 2 {
+		return fmt.Errorf("Was expecting WorkerPoolID to have syntax <provisionerId>/<workerType> but was %q", userData.WorkerPoolID)
+	}
+
+	c.ProvisionerID = wp[0]
+	c.WorkerType = wp[1]
+	c.WorkerGroup = userData.WorkerGroup
+	c.RootURL = userData.RootURL
+
+	// We need a worker manager client for fetching taskcluster credentials.
+	// Ensure auth is disabled in client, since we don't have credentials yet.
+	wm := c.WorkerManager()
+	wm.Authenticate = false
+	wm.Credentials = nil
+
+	workerIdentityProof, err := json.Marshal(providerType)
+	if err != nil {
+		return fmt.Errorf("Could not marshal provider type %#v: %v", providerType, err)
+	}
+
+	reg, err := wm.RegisterWorker(&tcworkermanager.RegisterWorkerRequest{
+		WorkerPoolID:        userData.WorkerPoolID,
+		ProviderID:          userData.ProviderID,
+		WorkerGroup:         userData.WorkerGroup,
+		WorkerID:            c.WorkerID,
+		WorkerIdentityProof: json.RawMessage(workerIdentityProof),
+	})
+
+	if err != nil {
+		return fmt.Errorf("Could not register worker: %v", err)
+	}
+
+	c.AccessToken = reg.Credentials.AccessToken
+	c.Certificate = reg.Credentials.Certificate
+	c.ClientID = reg.Credentials.ClientID
+
+	// TODO: process reg.Expires
+
+	newestDeploymentID = func() (string, error) {
+		log.Print("Checking if there is a new deploymentId...")
+		wpfd, err := wm.WorkerPool(c.ProvisionerID + "/" + c.WorkerType)
+		if err != nil {
+			return "", fmt.Errorf("**** Can't reach worker-manager to see if there is a new deploymentId: %v", err)
+		}
+		workerManagerConfig := new(WorkerManagerConfig)
+		err = json.Unmarshal(wpfd.Config, &workerManagerConfig)
+		if err != nil {
+			return "", errors.New("WARNING: Can't decode /userData portion of worker type definition - probably somebody has botched a worker type update - not shutting down as in such a case, that would kill entire pool!")
+		}
+
+		publicHostSetup, err := workerManagerConfig.WorkerConfig.PublicHostSetup()
```

Comment:
Is this actually going to work with w-m? I don't see how the config as it is would work with this.

Reply:
Yes, it is tested via the CI of this PR itself, which is deployed with the code from this branch. It works the same way it worked with AWS Provisioner, the worker config contains a section for the generic-worker config file, but also supports providing other data for bootstrapping the worker.